### PR TITLE
[Core] Decrement lineage ref count of an actor when the actor task return object reference is deleted

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
```
std::vector<rpc::ObjectReference> TaskManager::AddPendingTask(..) {
   if (spec.IsActorTask()) {
    const auto actor_creation_return_id = spec.ActorCreationDummyObjectId();
    task_deps.push_back(actor_creation_return_id);
   }
}
```

When an actor method is called, the actor lineage ref count increments as it's seen as a dependency of the actor task.

However, when we know that the actor task won't be retried (i.e. the return object is out of scope), we forget to decrement the actor lineage ref count.

This PR fixes this so that we have a matching pair of increment and decrement.

Also removed some dead code.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/task_manager.cc [hunk 1]
```diff
@@ -1241,6 +1241,13 @@ int64_t TaskManager::RemoveLineageReference(const ObjectID &object_id,
       }
     }
 
+    if (it->second.spec.IsActorTask()) {
+      // We need to decrement the actor lineage ref count here
+      // since it's incremented during TaskManager::AddPendingTask.
+      const auto actor_creation_return_id = it->second.spec.ActorCreationDummyObjectId();
+      released_objects->push_back(actor_creation_return_id);
+    }
```
Comment (jjyao) [inline src/ray/core_worker/task_manager.cc:1249]: Actual fix

File: src/ray/gcs/pb_util.h [hunk 2]
```diff
@@ -82,33 +82,6 @@ inline std::shared_ptr<ray::rpc::ErrorTableData> CreateErrorTableData(
   return error_info_ptr;
 }
 
-/// Helper function to produce actor table data.
-inline std::shared_ptr<ray::rpc::ActorTableData> CreateActorTableData(
```
Comment (stephanie-wang) [inline src/ray/gcs/pb_util.h:86]: Was this just unused?

**Review by stephanie-wang — APPROVED** [review]

File: src/ray/gcs/pb_util.h (hunk 2 above)
Comment (jjyao) [reply src/ray/gcs/pb_util.h:86]: Yes, dead code

**Comment by farmer00317558** [comment]
@jjyao Hello, I would like to ask, what phenomenon will this issue cause? Will it only affect the scenario where the Actor fails and retries? Or could it potentially lead to memory leaks during the execution of the Actor tasks?

I want to determine whether an upgrade to the latest version is necessary. 

Thanks.

**Pull request merged**